zebra-state/src/service/write.rs

```diff
@@ -0,0 +1,130 @@
+        QueuedNonFinalized,
+    >,
+    mut finalized_state: FinalizedState,
+    invalid_block_reset_sender: tokio::sync::mpsc::UnboundedSender<block::Hash>,
+    chain_tip_sender: Arc<Mutex<ChainTipSender>>,
+) {
+    while let Some(ordered_block) = finalized_block_write_receiver.blocking_recv() {
+        // TODO: split these checks into separate functions
+
+        if invalid_block_reset_sender.is_closed() {
+            info!("StateService closed the block reset channel. Is Zebra shutting down?");
+            return;
+        }
+
+        // Discard any children of invalid blocks in the channel
+        let next_valid_height = finalized_state
+            .db
+            .finalized_tip_height()
+            .map(|height| (height + 1).expect("committed heights are valid"))
+            .unwrap_or(Height(0));
+
+        if ordered_block.0.height > next_valid_height {
+            debug!(
+                ?next_valid_height,
+                invalid_height = ?ordered_block.0.height,
+                invalid_hash = ?ordered_block.0.hash,
+                "got a block that was too high. \
+                 Assuming a parent block failed, and dropping this block",
+            );
+
+            // We don't want to send a reset here, because it could overwrite a valid sent hash
+        }
+
+        // Try committing the block
+        match finalized_state.commit_finalized(ordered_block) {
+            Ok(finalized) => {
+                let tip_block = ChainTipBlock::from(finalized);
+
+                // TODO: update the chain tip sender with non-finalized blocks in this function,
+                //       and get rid of the mutex
+                chain_tip_sender
+                    .lock()
+                    .expect("unexpected panic in block commit task or state")
+                    .set_finalized_tip(tip_block);
+            }
+            Err(error) => {
+                let finalized_tip = finalized_state.db.tip();
+
+                // The last block in the queue failed, so we can't commit the next block.
+                // Instead, we need to reset the state queue,
+                // and discard any children of the invalid block in the channel.
+                info!(
+                    ?error,
+                    last_valid_height = ?finalized_tip.map(|tip| tip.0),
+                    last_valid_hash = ?finalized_tip.map(|tip| tip.1),
+                    "committing a block to the finalized state failed, resetting state queue",
+                );
+
+                let send_result =
+                    invalid_block_reset_sender.send(finalized_state.db.finalized_tip_hash());
+
+                if send_result.is_err() {
+                    info!("StateService closed the block reset channel. Is Zebra shutting down?");
+                    return;
+                }
+            }
+        }
+    }
+
+    // We're finished receiving finalized blocks from the state.
+    finalized_block_write_receiver.close();
```

Comment:
```suggestion
    // We're finished receiving finalized blocks from the state.
```

we should get rid of this since it would only ever reach past the blocking_recv call once the channel is already closed and empty

Reply:
Thanks, fixed in 870fa3fae4